[quant][graphmode] Fold prepacked weight into module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26939 [quant][graphmode] Integrate prepacked workaround in QuantFusion
* **#26579 [quant][graphmode] Fold prepacked weight into module**

Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

jerryzh168 committed:
[quant][graphmode] Fold prepacked weight into module

Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1410,6 +1410,68 @@ def forward(self, x):
+        w, b = m._c._get_module('_packed_linear_weight_bias')._get_method('_weight_bias')()
+        self.assertEqual(ref_w, w)
+        self.assertEqual(ref_b, b)
+        self.assertEqual(ref_res, res)
```
Since this is primarily for serialization, we need to add a test that checks that.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -910,5 +910,48 @@ graph(%self, %scale, %zero_point, %dtype):
   rewriter.runOnGraph(graph, filter);
 }
+
+void FoldPrepackedWeightIntoModule(
```
Should we recursively run it on submodules?

jerryzh168 replied:
OK

ZolotukhinM reviewed (changes requested):
Overall looks good, but I have some remarks (see inline).

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
should work now

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -939,5 +939,103 @@ graph(%self, %scale, %zero_point, %dtype):
+        script::Module callee_module = module;
+        if (module_instance->node()->kind() == prim::GetAttr) {
+          auto child_module_name = module_instance->node()->s(attr::name);
+          auto child_module = module.find_module(child_module_name);
```
We should go through all submodules, there is no reason to scan IR for call-methods for that.

jerryzh168 replied:
you mean all methods of all submodules? what if there are some methods not used?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -939,5 +939,103 @@ graph(%self, %scale, %zero_point, %dtype):
+    auto w_quant_val = values_map.at(vmap.at("w_quant"));
+    auto packed_params_val = values_map.at(vmap.at("packed_params"));
+
+    WithInsertPoint ins(b_val->node());
```
This is unnecessary since you're using `insertAfter` anyway.

ZolotukhinM reviewed (changes requested):
How would it work in the end-to-end workflow? QuantFusion happens in the graph-executor, and this should happen before. How would it find any `quantized::linear_prepack`?

jerryzh168 commented:
> How would it work in the end-to-end workflow? QuantFusion happens in the graph-executor, and this should happen before. How would it find any `quantized::linear_prepack`?

QuantFusion should happen before this, then maybe we should do quantfusion for linear and conv2d in advance?

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,74 @@ graph(%self, %scale, %zero_point, %dtype):
+      auto w = module.get_parameter("weight").variable_data();
+      auto w_quant = at::quantize_per_tensor(w, w_scale, w_zero_point, w_dtype);
+      auto b = module.get_parameter("bias").variable_data();
+      auto wrapper = wrapper_module.clone();
```
nice to see that the wrapped workaround works :)

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+
+      // m = self._packed_linear_weight_bias
+      Node* packed_params_module = graph->create(prim::GetAttr);
+      packed_params_module->addInput(graph->inputs()[0]);
```
Nit: you could combine these two lines into one:
```
Node* packed_params_module = graph->create(prim::GetAttr, {graph->inputs()[0]});
```
Also, you can chain some of these lines (e.g. `graph->create(...)->insertAfter(...)`).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      Node* packed_params_module = graph->create(prim::GetAttr);
+      packed_params_module->addInput(graph->inputs()[0]);
+      packed_params_module->s_(attr::name, "_packed_linear_weight_bias");
+      packed_params_module->output()->setDebugName("m");
```
Usually when we set a debug name, we get a name of an existing value in the graph and append it with some suffix. Setting names like "m" will screw up the existing names "m" which are pretty likely to be there from the original (user's) code. It's all only debug names, so everything would work correctly anyways, but I don't think we should destroy the existing names if we can avoid that. At least we should prepend the names of generated values with underscores or something to reduce chances of collision with real names.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      packed_params_module->insertAfter(w_quant_val->node());
+
+      // _packed_params = m._packed_params
+      Node* packed_params = graph->create(prim::GetAttr);
```
We should probably be using `Graph::insertGetAttr` for these.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      unpack_node->output()->setType(TupleType::create(std::vector<TypePtr>({TensorType::get(), OptionalType::create(TensorType::get())})));
+      unpack_node->insertAfter(packed_params);
+
+      WithInsertPoint ins(unpack_node);
```
If we're using the insert point API, let's use it everywhere for consistency: please replace `insertAfter` with `insertNode`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      // weight = weight_bias[0]
+      Node* weight_node = graph->create(prim::TupleIndex);
+      weight_node->addInput(unpack_node->output());
+      weight_node->addInput(index_zero);
```
Nit: please specify inputs in the `create` call.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      // bias = weight_bias[1]
+      Node* bias_node = graph->create(prim::TupleIndex);
+      bias_node->addInput(unpack_node->output());
+      bias_node->addInput(index_one);
```
Same.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -886,5 +886,131 @@ graph(%self, %scale, %zero_point, %dtype):
+      b_val->replaceAllUsesWith(bias_node->output());
+
+      // Delete nodes
+      std::vector<Node*> nodes_to_delete;
```
Nit:
```
std::vector<Node*> nodes_to_delete({w_val->node(), b_val->node(), w_quant_val->node()});
```

ZolotukhinM reviewed (changes requested):
I think this pass should be split into two parts: 1) insert pack-unpack nodes, 2) fold pack-attrs. Probably, the first part should actually be a part of insert quant-dequant, as it's semantically tied to that pass and implementation is (almost) the same. Such separation will later help us to generalize the second step into "fold whatever computations we can into attributes that are marked as constants", To make it work, I think we'll need to perform a transformation like the following:

Original:
```
%y = linear(%x, %w, %b)
```
After insert q-dq and pack/unpack:
```
%x_dq = dequant(quant(%x))
%packed = prepack(%w, %b)
SetAttr["_packed"](%packed)
%packed = GetAttr["_packed"]
%w_unpacked, %b_unpacked = unpack(%packed)
%y = linear(%x_dq, %w_unpacked, %b_unpacked)
%y_dq = dequant(quant(%y))
```

Then folding pass (the part 2 in my terminology) will look for patterns like
```
%w = GetAttr["weight"]
%b = GetAttr["bias"]
%packed = prepack(%w, %b)
SetAttr["_packed"](%packed)
```
It will precompute attribute "_packed" and remove the prepack node from IR (disclaimer: it will probably be more complicated than just searching for a pattern because get-attr might be a different function, but we're already dealing with that in quantization). Note that later we can precompute anything that operates on constant attributes at this step, not only prepack.

And ultimately, the fusion would fuse
```
%x_dq = dequant(%xq)
%w_unpacked, %b_unpacked = unpack(%packed)
%y = linear(%x_dq, %w_unpacked, %b_unpacked)
%yq = quant(%y)
```
with a quantized linear that takes packed weight and bias.

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
sounds reasonable, but inserting pack/unpack will need to be a separate pass since there is no generic pack/unpack(we have linear_prepack and conv2d_prepack) but quant/dequant is generic. So it's better to keep them separate.

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -887,5 +887,86 @@ void InsertPackUnpack(script::Module& module) {
+    script::parseIR(pattern, &pattern_graph, vmap);
+    const auto& matches = findPatternMatches(pattern_graph, *graph);
+    TORCH_CHECK(
+        matches.size() <= 1, "We only support at most one match right now");
```
This should not be an assertion failure. We should either support this or gracefully bail out.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -127,5 +127,24 @@ TORCH_API void InsertPackUnpack(script::Module& module);
+/** \brief Fold prepack function call into module
+ *
+ *  For the graph of the specified method, if we find a `quantized::linear_prepack` call,
+ *   we'll clone the wrapper module and set the weight
```
Nit: formatting :)

ZolotukhinM reviewed (approved):
Looks good! Some nits are inline.

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Fold prepacked weight into module"


Summary:
Remove `linear_prepack` call and attach a module to the
parent class that contains the packed weight and bias,
this is to support serialization of the quantized model
since the packed weight and bias is not serializable and
we need to overwrite the `__getstate__` and `__setstate__`
function to be able to serialize them

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636397](https://our.internmc.facebook.com/intern/diff/D17636397)

[ghstack-poisoned]
